File: src/infrastructure/external/hko_client.go
// Package external integrates with Hong Kong-specific public services that
// turn this booking system into a localized product:
//
//   - hko_client.go        HKO weather warning feed → triggers exception flow
//   - govhk_holidays.go    gov.hk public-holiday ICS feed → auto-import
//
// Both are best-effort enrichments. The system continues to work without
// them; missing data is logged but never fails a booking.
package external

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// HKO publishes a JSON warning summary at this stable endpoint.
const hkoWarningURL = "https://data.weather.gov.hk/weatherAPI/opendata/weather.php?dataType=warnsum&lang=en"

// WeatherSignal codes we react to. The HKO returns codes like "TC8NE",
// "TC10", "WRAINB" (Black Rainstorm). We collapse to a shape useful for
// the exception workflow.
type WeatherSignal struct {
	Code        string    // "T1" / "T3" / "T8" / "T10" / "BLACK_RAIN" / "RED_RAIN" / "AMBER_RAIN"
	Severity    int       // 1 (mild) … 10 (extreme)
	Description string
	IssuedAt    time.Time
}

// HKOClient queries the HKO warnings API.
type HKOClient struct {
	http *http.Client
}

func NewHKOClient(timeout time.Duration) *HKOClient {
	if timeout == 0 {
		timeout = 8 * time.Second
	}
	return &HKOClient{http: &http.Client{Timeout: timeout}}
}

// CurrentSignals fetches the active weather signals. The HKO response
// shape is a flat object whose keys are warning identifiers; we project
// only the ones we care about for booking exceptions.
//
// Reference response keys: WTCSGNL (typhoon), WRAIN (rainstorm).
func (c *HKOClient) CurrentSignals(ctx context.Context) ([]WeatherSignal, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, hkoWarningURL, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("hko: unexpected status %d", resp.StatusCode)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// Decode the heterogeneous map then extract the keys we care about.
	var raw map[string]struct {
		Name       string `json:"name"`
		Code       string `json:"code"`
		Type       string `json:"type"`
		ActionCode string `json:"actionCode"`
		IssueTime  string `json:"issueTime"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, err
	}

	var out []WeatherSignal
	for _, v := range raw {
		sig, ok := classify(v.Code, v.Name)
		if !ok {
			continue
		}
		issued, _ := time.Parse(time.RFC3339, v.IssueTime)
		sig.IssuedAt = issued
		out = append(out, sig)
	}
	return out, nil
}

// classify converts an HKO warning code to our internal taxonomy.
// Returns (zero, false) for warnings we don't act on (e.g. fire danger).
func classify(code, name string) (WeatherSignal, bool) {
	c := strings.ToUpper(code)
	n := strings.ToUpper(name)
	switch {
	case strings.Contains(c, "TC10"):
		return WeatherSignal{Code: "T10", Severity: 10, Description: name}, true
	case strings.Contains(c, "TC9"):
		return WeatherSignal{Code: "T9", Severity: 9, Description: name}, true
	case strings.Contains(c, "TC8"):
		return WeatherSignal{Code: "T8", Severity: 8, Description: name}, true
	case strings.Contains(c, "TC3"):
		return WeatherSignal{Code: "T3", Severity: 3, Description: name}, true
	case strings.Contains(c, "TC1"):
		return WeatherSignal{Code: "T1", Severity: 1, Description: name}, true
	case c == "WRAINB" || strings.Contains(n, "BLACK"):
		return WeatherSignal{Code: "BLACK_RAIN", Severity: 9, Description: name}, true
	case c == "WRAINR" || strings.Contains(n, "RED"):
		return WeatherSignal{Code: "RED_RAIN", Severity: 7, Description: name}, true
	case c == "WRAINA" || strings.Contains(n, "AMBER"):
		return WeatherSignal{Code: "AMBER_RAIN", Severity: 4, Description: name}, true
	}
	return WeatherSignal{}, false
}

// SuspendsBookings reports whether a signal is severe enough that no-show
// penalties should be auto-suspended (T8+, Black Rain).
func (s WeatherSignal) SuspendsBookings() bool {
	return s.Severity >= 8
}

File: fsd-v2/backend/src/modules/weather/hko.client.ts
import { Injectable, Logger } from '@nestjs/common';
import axios from 'axios';

// HKO open-data endpoints — stable URLs (https://data.weather.gov.hk).
const HKO_WARN_URL = 'https://data.weather.gov.hk/weatherAPI/opendata/weather.php?dataType=warnsum&lang=en';
const HKO_CURRENT_URL = 'https://data.weather.gov.hk/weatherAPI/opendata/weather.php?dataType=rhrread&lang=en';

export interface WeatherSignal {
  code: string;         // 'T1' | 'T3' | 'T8' | 'T10' | 'BLACK_RAIN' | …
  severity: number;     // 1 (mild) … 10 (extreme)
  description: string;
  issuedAt?: string;    // ISO
}

export interface WeatherReport {
  tempC: number;
  signals: WeatherSignal[];
  updatedAt: string;
}

@Injectable()
export class HKOClient {
  private readonly log = new Logger(HKOClient.name);

  async currentWeather(): Promise<WeatherReport> {
    const [signals, current] = await Promise.allSettled([
      this.currentSignals(),
      axios.get(HKO_CURRENT_URL, { timeout: 8_000 }),
    ]);

    const out: WeatherReport = {
      tempC: 0,
      signals: signals.status === 'fulfilled' ? signals.value : [],
      updatedAt: new Date().toISOString(),
    };

    if (current.status === 'fulfilled') {
      const raw = current.value.data as any;
      // The hottest reading across stations is the safety-relevant one.
      for (const d of raw?.temperature?.data ?? []) {
        if (typeof d.value === 'number' && d.value > out.tempC) out.tempC = d.value;
      }
      if (raw?.temperature?.recordTime) out.updatedAt = raw.temperature.recordTime;
    }
    return out;
  }

  async currentSignals(): Promise<WeatherSignal[]> {
    const res = await axios.get(HKO_WARN_URL, { timeout: 8_000 });
    const raw = (res.data ?? {}) as Record<string, { name?: string; code?: string; issueTime?: string }>;
    const out: WeatherSignal[] = [];
    for (const v of Object.values(raw)) {
      const sig = this.classify(v.code ?? '', v.name ?? '');
      if (!sig) continue;
      out.push({ ...sig, issuedAt: v.issueTime });
    }
    return out;
  }

  // Internal taxonomy — we collapse HKO codes into a shape useful for
  // the booking exception workflow (T8+ / Black Rain auto-suspends bookings).
  private classify(code: string, name: string): WeatherSignal | null {
    const c = code.toUpperCase();
    const n = name.toUpperCase();
    if (c.includes('TC10')) return { code: 'T10', severity: 10, description: name };
    if (c.includes('TC9'))  return { code: 'T9',  severity: 9,  description: name };
    if (c.includes('TC8'))  return { code: 'T8',  severity: 8,  description: name };
    if (c.includes('TC3'))  return { code: 'T3',  severity: 3,  description: name };
    if (c.includes('TC1'))  return { code: 'T1',  severity: 1,  description: name };
    if (c === 'WRAINB' || n.includes('BLACK'))  return { code: 'BLACK_RAIN', severity: 9, description: name };
    if (c === 'WRAINR' || n.includes('RED'))    return { code: 'RED_RAIN',   severity: 7, description: name };
    if (c === 'WRAINA' || n.includes('AMBER'))  return { code: 'AMBER_RAIN', severity: 4, description: name };
    return null;
  }
}

// Helper exported for callers that want the booking-suspension rule
// without re-implementing the threshold.
export function suspendsBookings(s: WeatherSignal): boolean {
  return s.severity >= 8;
}
